src/ControlSystem/TimescaleTuner.cpp

```diff
@@ -0,0 +1,126 @@
+      max_threshold_{max_threshold},
+      min_threshold_{min_threshold},
+      increase_factor_{increase_factor},
+      decrease_factor_{decrease_factor} {
+  for (auto t_scale : timescale_) {
+    if (t_scale <= 0.0) {
+      ERROR("Initial timescale must be > 0");
+    }
+  }
+
+  if (decrease_factor_ > 1.0 or decrease_factor <= 0.0) {
+    ERROR("The specified decrease_factor "
+          << decrease_factor_ << " must satisfy 0 < decrease_factor <= 1");
+  }
+  if (increase_factor_ < 1.0) {
+    ERROR("The specified increase factor " << increase_factor_
+                                           << " must be >= 1.0");
+  }
+  if (min_timescale_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_timescale_ <= min_timescale_) {
+    ERROR("The maximum timescale "
+          << max_timescale_
+          << " must be > than the specified minimum timescale "
+          << min_timescale_);
+  }
+  if (min_threshold_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_threshold_ <= min_threshold_) {
+    ERROR("The maximum threshold "
+          << max_threshold_
+          << " must be > than the specified minimum threshold "
+          << min_threshold_);
+  }
+}
+
+void TimescaleTuner::suggested_timescale_if_in_allowable_range(
+    double suggested_timescale) noexcept {
+  for (auto& t_scale : timescale_) {
+    if (suggested_timescale > max_timescale_) {
+      t_scale = max_timescale_;
+    } else if (suggested_timescale < min_timescale_) {
+      t_scale = min_timescale_;
+    } else {
+      t_scale = suggested_timescale;
+    }
+  }
+}
+
+void TimescaleTuner::update_timescale(
+    const std::array<DataVector, 3>& qs) noexcept {
+  // above, Q is an array of length 3: \int Q, Q, dQ/dt
+  // each of these 3 is a DataVector w/ length = the number of components
+
+  if (qs[0].size() != timescale_.size()) {
+    ERROR("The number of components in qs("
+          << qs[0].size() << ") does not match the number of timescales("
+          << timescale_.size() << ")");
+  }
+  const DataVector& q = gsl::at(qs, 1);
+  const DataVector& dtq = gsl::at(qs, 2);
+
+  for (size_t i = 0; i < q.size(); i++) {
+    bool increase_timescale = false;
+    bool decrease_timescale = false;
+
+    // check whether we need to decrease the timescale:
+    // if Q or dtQ are above the maximum tolerance
+    if (fabs(q[i]) > max_threshold_ ||
+        fabs(dtq[i] * timescale_[i]) > max_threshold_) {
+      // AND
+      // (1) Q and dtQ are the same sign: the error is growing
+      // or (2) if Q is not expected to drop to half of its current value in
+      // one timescale: not decreasing fast enough
+      if (dtq[i] * q[i] > 0.0 ||
+          fabs(dtq[i]) * timescale_[i] < 0.5 * fabs(q[i])) {
+        decrease_timescale = true;
+      }
```

Comment:
If you set decrease_timescale = true, can't you then skip the checks for increase_timescale?

Reply:
I.e., maybe instead of setting increase_timescale and decrease_timescale, just do the increase/decrease and return? As long as you check for decreases first, shouldn't that be fine?

Reply:
I've removed the unnecessary bool flags. Those were from SpEC, where a check was done to see if both increase and decrease triggered, but I think I found it impossible to reach both, so I removed that check (but overlooked their uselessness beyond that check). I moved the increase/decrease into the conditionals, but left the minimum/maximum check at the bottom, since it is common to both.